python/paddle/fluid/dygraph/dygraph_to_static/cache_program.py

```diff
@@ -0,0 +1,292 @@
+
+        if last_func == self.forward_func and not self.is_build_process:
+            outputs = self._run(args, kwargs)
+
+        return outputs
+
+    def _check_cache_valid(self):
+        """
+        Checks whether the current program is consistent with `default_main_program`.
+        In some models and unittest, program will be switched frequently by `program_guard`.
+        If does, the cached program and other properties are not available and should be reset.
+        """
+        if self.program:
+            if self.program != framework.default_main_program():
+                ProgramCache.reset()
+        # Always set program to default_main_program. Because once `__call__` is called,
+        # it means layers(or Ops) are added into default_main_program switched by outer
+        # `with` statement.
+        self.program = framework.default_main_program()
+
+    def _prepare(self, args, kwargs):
+        """
+        Prepares batch_data and adds `fluid.data` layers into program.
+        """
+        if not self.inputs:
+            self._add_feed_layers(args, kwargs)
+        elif self.traced_funcs[-1] == self.forward_func:
+            self._update_batch_data(args, kwargs)
+
+    def add_layers(self, dyfunc):
+        """
+        Transforms dygraph function into static function, and adds layers into
+        main_program.
+        """
+        self._check_cache_valid()
+        static_func = self.func_cache(dyfunc)
+        # self._forward_func is entry function of Net or Model.
+        # It can be called for multiple times, but layers from these functions
+        # call stack will be added into self._program only once.
+        # After that, cached program will be always returned by default.
+        if static_func == self.forward_func:
+            self.is_repeated = True
+
+        if self.forward_func is None:
+            self.forward_func = static_func
+
+        self.traced_funcs.append(static_func)
+
+        return static_func
+
+    def _get_or_build_program(self, func, args, kwargs):
+        """
+        Returns program of the input function. If called at first time,
+        builds a new program and caches it.
+        """
+        with framework.program_guard(self.program):
+            if func == self.forward_func and self.is_build_process:
+                # Replaces input data with `layers.data`
+                feed_name_to_idx = self._feed_name_to_idx(self.forward_func)
+                args = list(args)
+                for feed_layer in self.inputs:
+                    idx = feed_name_to_idx[feed_layer.name]
+                    args[idx] = feed_layer
+                self.is_build_process = False
+                fetch_list = func(*args, **kwargs)
+                self._add_optimizer(fetch_list)
+            else:
+                fetch_list = func(*args, **kwargs)
+        self.outputs = fetch_list
+        return fetch_list
+
+    def _run(self, args, kwargs):
```

Comment:
As we have discussed, the run logic had better to be extract out. Let's make the cache only does cache thing, check valid, query and get, reset. Move the logic of adding/modifying programs to somewhere else. That's Object Oriented Programming (OOP) :-)

Then it will increase readability and people with cache background knowledge can read it easier.

Reply:
Great advice. Fixed.